{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'gene_finder'",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mModuleNotFoundError\u001b[39m                       Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[16]\u001b[39m\u001b[32m, line 2\u001b[39m\n\u001b[32m      1\u001b[39m \u001b[38;5;66;03m#| hide\u001b[39;00m\n\u001b[32m----> \u001b[39m\u001b[32m2\u001b[39m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01mgene_finder\u001b[39;00m\u001b[34;01m.\u001b[39;00m\u001b[34;01mcore\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m *\n",
      "\u001b[31mModuleNotFoundError\u001b[39m: No module named 'gene_finder'"
     ]
    }
   ],
   "source": [
    "#| hide\n",
    "from gene_finder.core import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# gene_finder\n",
    "> Utility to identify gene sequences from WGS data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Quick Gene Assembly and Annotation Pipeline\n",
    "\n",
    "- We have several short-read WGS sequencing datasets from our mode of action studies. \n",
    "Sometimes we need to retrieve the full length of a coding sequence when it's not \n",
    "well-annotated in public databases.\n",
    "- We can leverage **SPAdes** for de novo assembly \n",
    "and **MMseqs2** for rapid gene finding to quickly identify and extract complete \n",
    "gene sequences from our sequencing data.\n",
    "\n",
    "### Installation\n",
    "\n",
    "First, install the required tools using conda:\n",
    "```bash\n",
    "conda install -c conda-forge -c bioconda mmseqs2 spades\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/mtinti/git_projects/gene_finder/data/V350230622_L01_B5GANIdzjaRAAAA-1_R1.fastq.gz\n",
      "/Users/mtinti/git_projects/gene_finder/data/V350230622_L01_B5GANIdzjaRAAAA-1_R2.fastq.gz\n",
      "/Users/mtinti/git_projects/gene_finder/data/V350370689_L02_B5GANIokjbRAAAA-128_1.fq.gz\n",
      "/Users/mtinti/git_projects/gene_finder/data/V350370689_L02_B5GANIokjbRAAAA-128_2.fq.gz\n",
      "/Users/mtinti/git_projects/gene_finder/data/V350370689_L03_B5GANIokjbRAAAA-128_1.fq.gz\n",
      "/Users/mtinti/git_projects/gene_finder/data/V350370689_L03_B5GANIokjbRAAAA-128_2.fq.gz\n",
      "\n",
      "/Users/mtinti/git_projects/gene_finder/data/output_Tc_Sylvio:\n",
      "assembly_graph_after_simplification.gfa \u001b[34mK77\u001b[m\u001b[m\n",
      "assembly_graph_with_scaffolds.gfa       \u001b[34mmisc\u001b[m\u001b[m\n",
      "assembly_graph.fastg                    params.txt\n",
      "before_rr.fasta                         \u001b[34mpipeline_state\u001b[m\u001b[m\n",
      "contigs.fasta                           run_spades.sh\n",
      "contigs.paths                           run_spades.yaml\n",
      "\u001b[34mcorrected\u001b[m\u001b[m                               scaffolds.fasta\n",
      "dataset.info                            scaffolds.paths\n",
      "input_dataset.yaml                      spades.log\n",
      "\u001b[34mK21\u001b[m\u001b[m                                     \u001b[34mtmp\u001b[m\u001b[m\n",
      "\u001b[34mK33\u001b[m\u001b[m                                     warnings.log\n",
      "\u001b[34mK55\u001b[m\u001b[m\n"
     ]
    }
   ],
   "source": [
    "# our sequence libraries\n",
    "!ls /Users/mtinti/git_projects/gene_finder/data/*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Running Spades\n",
    "\n",
    "This is the most time comsing part, this dataset run for 1h approx\n",
    "```bash\n",
    "spades.py -t 100 \\\n",
    "--pe1-1 ../data/V350230622_L01_B5GANIdzjaRAAAA-1_R1.fastq.gz \\\n",
    "--pe1-2 ../data/V350230622_L01_B5GANIdzjaRAAAA-1_R2.fastq.gz \\\n",
    "--pe2-1 ../data/V350370689_L02_B5GANIokjbRAAAA-128_1.fq.gz \\\n",
    "--pe2-2 ../data/V350370689_L02_B5GANIokjbRAAAA-128_2.fq.gz \\\n",
    "--pe3-1 ../data/V350370689_L03_B5GANIokjbRAAAA-128_1.fq.gz \\\n",
    "--pe3-2 ../data/V350370689_L03_B5GANIokjbRAAAA-128_2.fq.gz \\\n",
    "-o ../data/output_Tc_Sylvio\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Running mmseq2\n",
    "Now we use mmseq2 to look for our gene on interest in the assebled scaffolds\n",
    "```bash\n",
    "!mmseqs easy-search --search-type 3 \\\n",
    "--format-output  query,evalue,qstart,qend,qlen,target,tstart,tend \\\n",
    "../data/CLK1_CDS_Dm28c.fa ../data/output_Tc_Sylvio/scaffolds.fasta \\\n",
    "../data/out.CLK1.tsv ../data/tmp\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CLK1\t0.000E+00\t369\t1458\t1458\tNODE_2457_length_1097_cov_126.120588\t1\t1090\n",
      "CLK1\t2.871E-225\t1\t445\t1458\tNODE_289_length_20803_cov_57.477757\t20359\t20803\n",
      "CLK1\t1.021E-38\t445\t354\t1458\tNODE_2388_length_1130_cov_56.830959\t1\t92\n",
      "CLK1\t1.716E-25\t1389\t1458\t1458\tNODE_1614_length_1713_cov_44.015892\t1\t70\n",
      "CLK1\t1.716E-25\t1389\t1458\t1458\tNODE_55_length_75916_cov_61.009375\t1\t70\n"
     ]
    }
   ],
   "source": [
    "#peek into the output\n",
    "!head ../data/out.CLK1.tsv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- The CLK1 sequence is split into two contigs\n",
    "\n",
    "Let's retrive the sequences +/- 1000 bp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid syntax (2504544915.py, line 1)",
     "output_type": "error",
     "traceback": [
      "  \u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[19]\u001b[39m\u001b[32m, line 1\u001b[39m\n\u001b[31m    \u001b[39m\u001b[31mfrom\u001b[39m\n         ^\n\u001b[31mSyntaxError\u001b[39m\u001b[31m:\u001b[39m invalid syntax\n"
     ]
    }
   ],
   "source": [
    "from \n",
    "retrieve_and_save_sequence('../data/output_Tc_Sylvio/scaffolds.fasta',\n",
    "                           'NODE_2457_length_1097_cov_126.120588',\n",
    "                           1,1090+1000,'../data/out_NODE_2457.fa')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Developer Guide"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If you are new to using `nbdev` here are some useful pointers to get you started."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Install gene_finder in Development mode"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```sh\n",
    "# make sure gene_finder package is installed in development mode\n",
    "$ pip install -e .\n",
    "\n",
    "# make changes under nbs/ directory\n",
    "# ...\n",
    "\n",
    "# compile to have changes apply to gene_finder\n",
    "$ nbdev_prepare\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Installation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Install latest from the GitHub [repository][repo]:\n",
    "\n",
    "```sh\n",
    "$ pip install git+https://github.com/mtinti/gene_finder.git\n",
    "```\n",
    "\n",
    "or from [conda][conda]\n",
    "\n",
    "```sh\n",
    "$ conda install -c mtinti gene_finder\n",
    "```\n",
    "\n",
    "or from [pypi][pypi]\n",
    "\n",
    "\n",
    "```sh\n",
    "$ pip install gene_finder\n",
    "```\n",
    "\n",
    "\n",
    "[repo]: https://github.com/mtinti/gene_finder\n",
    "[docs]: https://mtinti.github.io/gene_finder/\n",
    "[pypi]: https://pypi.org/project/gene_finder/\n",
    "[conda]: https://anaconda.org/mtinti/gene_finder"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Documentation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Documentation can be found hosted on this GitHub [repository][repo]'s [pages][docs]. Additionally you can find package manager specific guidelines on [conda][conda] and [pypi][pypi] respectively.\n",
    "\n",
    "[repo]: https://github.com/mtinti/gene_finder\n",
    "[docs]: https://mtinti.github.io/gene_finder/\n",
    "[pypi]: https://pypi.org/project/gene_finder/\n",
    "[conda]: https://anaconda.org/mtinti/gene_finder"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## How to use"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Fill me in please! Don't forget code examples:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "1+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python3",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
